### PR TITLE
feat(examples) demonstrate building nextjs app

### PR DESCRIPTION
**Pull request by ewianda**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature (example)
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: N/A

## What is the new behavior?
Example demonstration how to build a nextjs application

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by alexeagle** [comment]
Cool, do you want feedback on this PR and make edits yourself, to learn more Bazel stuff? Or would you like me to just fix it up?

**Comment by ewianda** [comment]
> Cool, do you want feedback on this PR and make edits yourself, to learn more Bazel stuff? Or would you like me to just fix it up?

Will love to lean more stuff :grin:

File: examples/nextjs/pages/_app.tsx [hunk 1]
```diff
@@ -0,0 +1,6 @@
+import { AppProps } from 'next/app'
+import '../styles/index.css'
```
Comment (alan97) [inline examples/nextjs/pages/_app.tsx:2]: Should global.css also be imported here? I am curious to see if with your example you can reproduce https://github.com/vercel/next.js/issues/15218

**Review by alan97 — COMMENTED** [review]
I am encountering this issue https://github.com/vercel/next.js/issues/15218 with Tailwind 3 + Next.js 10 when running the build process with Bazel. I am curious if this issue still persists with Next.js 12 + Tailwind 3 and if you can include the global CSS import as part of the example. Thanks for putting this together by the way!

**Comment by ewianda** [comment]
@alexeagle just a friendly reminder on the review.

**Comment by supercairos** [comment]
I'm currious, why are all the files wrapped as `ts_project` ?

**Comment by ewianda** [comment]
> I'm currious, why are all the files wrapped as `ts_project` ?

ts_project is the [recommended](https://bazelbuild.github.io/rules_nodejs/TypeScript.html) way of building typescript code

**Comment by supercairos** [comment]
> > I'm currious, why are all the files wrapped as `ts_project` ?
> 
> ts_project is the [recommended](https://bazelbuild.github.io/rules_nodejs/TypeScript.html) way of building typescript code

I understand that, but why not a big ts_project with all the file in-it ? It's easlily maintainable IMO

**Comment by ewianda** [comment]
> > > I'm currious, why are all the files wrapped as `ts_project` ?
> > 
> > 
> > ts_project is the [recommended](https://bazelbuild.github.io/rules_nodejs/TypeScript.html) way of building typescript code
> 
> I understand that, but why not a big ts_project with all the file in-it ? It's easlily maintainable IMO

At scale a big ts_project doesn't work because you lose the ability to utilise caching and even to parallise your builds with RBE, since a single file change means you have to rebuild everything in CI.  Also fine-grained dependencies is[ bazel's best practice](https://bazel.build/docs/best-practices)

**Comment by alexeagle** [comment]
I have some potential clients looking into next.js with Bazel, so it would be nice to get this in.

Re. the discussion of granularity of `ts_project` targets: there is a tradeoff. If you have a BUILD file generator like Gazelle, then there's not much expense for developers to tell Bazel the dependency graph, and you can afford to have a lot of those files. However there isn't a canonical Gazelle TS extension yet (Aspect is working on one: https://github.com/aspect-build/rules_js/pull/21).
As you say, a single `ts_project` for the entire application plus libraries doesn't deliver on Bazel's promise of faster builds through incrementality.
So a smaller number of "libraries", each with a `ts_project` rule, is a pretty good compromise.

**Comment by ewianda** [comment]
We opened source our[ Gazelle plugin](https://github.com/benchsci/rules_nodejs_gazelle) that works well for both TS and JS. We have two options, fine grain ( per file target) or aggregated ( per package targes). But we have never tried per package at scale.

**Comment by alexeagle** [comment]
Hey @ewianda - have any interest in picking this up again with me? I'm working over here on it 👉🏻 https://github.com/aspect-build/bazel-examples/tree/next/next.js
with the goal of not just calling the `next` tool to wrap their build system, but fully replace their build system with Bazel.

**Comment by ewianda** [comment]
> Hey @ewianda - have any interest in picking this up again with me? I'm working over here on it 👉🏻 https://github.com/aspect-build/bazel-examples/tree/next/next.js with the goal of not just calling the `next` tool to wrap their build system, but fully replace their build system with Bazel.

Yes i am interested

**Comment by alexeagle** [comment]
We have a Bazel + next.js example here now https://github.com/aspect-build/bazel-examples/tree/main/next.js

@ewianda sorry we didn't get this one merged, but I'd still love to have more of your contributions in the future! That example probably needs more polish and features.

**Pull request closed without merging**